src/itemmenu.cpp

```diff
@@ -952,15 +808,110 @@ static void sort_items()
  */
 int useup_item(int item_id)
 {
-    size_t inventory_index;
+    return g_inv.remove(item_id) ? 1 : 0;
+}
+
+// KInventory implementation
+
+s_inventory KInventory::operator[](int i)
+{
+    if (i >= 0 && i < inv.size())
+    {
+        return inv.at(i);
+    }
+    else
+    {
+        return {};
+    }
+}
 
-    for (inventory_index = 0; inventory_index < MAX_INV; ++inventory_index)
+void KInventory::add(int type, int quantity)
+{
+    inv.emplace_back(type, quantity);
+    normalize();
+}
+bool KInventory::remove(int type, int quantity)
+{
+    for (auto& it : inv)
     {
-        if (g_inv[inventory_index].item == item_id)
+        if (it.item == type)
         {
-            remove_item(inventory_index, 1);
-            return 1;
+            if (it.quantity >= quantity)
+            {
+                it.quantity -= quantity;
+                quantity = 0;
+            }
+            else
+            {
+                quantity -= it.quantity;
+                it.quantity = 0;
+            }
+            if (quantity == 0)
+            {
+                break;
+            }
         }
     }
-    return 0;
+    return quantity == 0;
+}
+bool KInventory::removeIndex(int ix, int quantity)
+{
+    if (ix >= 0 && ix < inv.size())
+    {
+        auto& it = inv.at(ix);
+        if (it.quantity >= quantity)
+        {
+            it.quantity -= quantity;
+            return true;
+        }
+        else
```

Comment:
NIT only: If there's only one place this can return true, the remaining `else`s can probably be omitted.

Then again, it doesn't hurt to keep it in to make it more explicit.

Reply:
I'll change it.